### ISSUE TITLE
Adding records to a one2many deletes records if the domain uses a lambda

### ISSUE THREAD

**Issue by Yajo** [odoo/odoo#18438]
Impacted versions: 8, 9, not tested in others

Steps to reproduce:

Install [`l10n_es_aeat_mod303`](https://www.odoo.com/apps/modules/9.0/l10n_es_aeat_mod303/) and [`l10n_es_aeat_mod111`](https://www.odoo.com/apps/modules/9.0/l10n_es_aeat_mod111/).

Generate one 303 report and one 111 report. You should ensure both have the same ID to reproduce the bug.

Calculate taxes for one and then for the other.

Current behavior: When the 2nd model generates its taxes, it deletes 1st's.

The offending line is [this one](https://github.com/OCA/l10n-spain/blob/0dfe806dcfddf595d1ab36ee1cba32d19156eb11/l10n_es_aeat/models/l10n_es_aeat_report_tax_mapping.py#L38). See the code:

```python
class L10nEsAeatReportTaxMapping(models.AbstractModel):
    _name = "l10n.es.aeat.report.tax.mapping"
    _inherit = "l10n.es.aeat.report"
    _description = ("Inheritable abstract model to add taxes by code mapping "
                    "in any AEAT report")

    tax_line_ids = fields.One2many(
        comodel_name='l10n.es.aeat.tax.line', inverse_name='res_id',
        domain=lambda self: [("model", "=", self._name)], auto_join=True,
        readonly=True, oldname='tax_lines', string="Tax lines")

    @api.multi
    def calculate(self):
        res = super(L10nEsAeatReportTaxMapping, self).calculate()
        for report in self:
            report.tax_line_ids.unlink()
            # Buscar configuración de mapeo de impuestos
            tax_code_map = self.env['l10n.es.aeat.map.tax'].search(
                [('model', '=', report.number),
                 '|',
                 ('date_from', '<=', report.date_start),
                 ('date_from', '=', False),
                 '|',
                 ('date_to', '>=', report.date_end),
                 ('date_to', '=', False)], limit=1)
            if tax_code_map:
                tax_lines = []
                for map_line in tax_code_map.map_line_ids:
                    tax_lines.append(report._prepare_tax_line_vals(map_line))
                report.tax_line_ids = [(0, 0, x) for x in tax_lines]
        return res
```

Notice that `tax_line_ids` is declared with `domain=lambda self: [("model", "=", self._name)]`, a common pattern when you are doing `one2many` to a flexible model.

Then, later, when doing `report.tax_line_ids = [(0, 0, x) for x in tax_lines]`, it produces an `unlink` of all `l10n.es.aeat.tax.line` where their id match to `report.id` but their `model` does not match the `report._name`.

We tried replacing the offending line with:

```python
for line in tax_lines:
    report.tax_line_ids |= self.env['l10n.es.aeat.tax.line'].new(line)
```

... but it happens the same.

Expected behavior: Adding records should not delete others.

@Tecnativa @rco-odoo

**Comment by Yajo** [comment]
A test for this has been added in #18440.

**Comment by pedrobaeza** [comment]
Any news on this?

**Comment by nim-odoo** [comment]
The fix is ongoing, see PR #19181. I added your test (which I had to slightly modify). Please note that we won't fix for 8.0: this version won't be supported anymore in a few weeks, so we don't want to take the risk to introduce an issue this late in its lifecycle.

**Comment by pedrobaeza** [comment]
Thanks for the update. 

We did a workaround meanwhile in the 3 versions (example for v10: https://github.com/OCA/l10n-spain/commit/35362e2ae9205851adc8f37e6acd6daee7380148). If v8 is not solved, we will keep the workaround any way. The important thing is to have it solved in newer versions.
